File: Week3/Code/Vectorize1.R
# Amy Solman amy.solman19@imperial.ac.uk
# 17th October 2019
# Vectorize1.R

# Creates a matrix (collection of elements of same data type -
# here numeric - arranged in fixed number of rows/columns.
# A 2D vector). Made of 1000000 random deviates, with 1000 rows and 1000 columns.
M <- matrix(runif(1000000),1000,1000) 

SumAllElements <- function(M){ # Create a function that will take the matrix
  # M as an argument
  Dimensions <- dim(M) # Dimensions of matrix applied to Dimensions variable
  Tot <- 0 # Numerical value 0 applied to 'Tot'-al
  for (i in 1:Dimensions[1]){ # For each element from 1 to the first element in dimensions (1000)
    for (j in 1:Dimensions[2]){ # For each element from 1 to the second element in dimensions (1000)
      Tot <- Tot + M[i,j] # Tot is assigned the value of Tot plus one element from first dimention 
      # and one element from second dimension of matrix
    }
  }
   return (Tot)             
}

print("Using loops, the time taken is:")
print(system.time(SumAllElements(M)))

print("Using the in-built vectorized function, the time taken is:")
print(system.time(sum(M)))

File: Week3/Code/Vectorize1.py
#!/usr/bin/env python3
# Date: Dec 2019

"""Creates a matrix (collection of elements of samer data type -
here numeric - arranged in fixed number of rows/columns.
A 2D vector, made of 1000000 random deviates, with 1000 rows and 1000 columns."""

__appname__ = 'Vectorize1.py'
__author__ = 'Amy Solman (amy.solman19@imperial.ac.uk)'
__version__ = '0.0.1'

import scipy as sc 
import numpy as np 
import time

#create a matrix of 1,000,000 random numbers (0,1) with 1000 x 1000 dimensions
M = np.random.uniform(size = 1000000)
M = M.reshape(1000,1000)
M = sc.matrix(M)

def SumAllElements(M):
    Dimensions = M.shape
    Sum = 0
    for i in range(Dimensions[0]):
        for j in range(Dimensions[1]):
            Sum = Sum + M[i,j]
    return(Sum)

start = time.time()
SumAllElements(M)
print("The loop function takes %fs to run." % (time.time() - start))

start = time.time()
np.sum(M)
print("The in-built vectorized function takes %fs to run." % (time.time() - start))



